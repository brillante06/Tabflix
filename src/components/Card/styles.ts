import styled from 'styled-components';

interface image {
    src: string;
    alt?: string;
}
export const movieImage = styled.img<image>`
    src: ${(props) => props.src};
    alt: ${(props) => props.alt};
    border-radius: 1rem;
`;
export const movieInfo = styled.p`
    font-size: 1.2rem;
    padding: -1rem 0 0 4rem;
    margin: 0.6rem 0 1rem 0.8rem;
`;
export const movieRating = styled.p`
    font-size: 1.2rem;
    padding: -1rem 0 0 4rem;
    margin: 0.4rem 0.8rem 1rem 0;
`;
export const Container = styled.article`
    border: 1px solid #ccc;
    border-radius: 1rem;
    box-shadow: 2px 2px 6px 0px rgba(0, 0, 0, 0.3);
    width: 10rem;
    height: 24rem;
`;
export const InfoContainer = styled.article`
    display: flex;
    justify-content: space-between;
`;
