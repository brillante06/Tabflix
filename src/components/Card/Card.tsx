import React, { FC } from 'react';
import { movieInfo } from '../../types';
import * as S from './styles';

interface movieCard {
    image: string;
    title: string;
    onClick: (movie: movieInfo, id: number) => void;
    id: number;
    movie: movieInfo;
}

const Card: FC<movieCard> = ({ image, title, onClick, id, movie }) => {
    const onClickMove = () => {
        onClick(movie, id);
    };
    return (
        <S.Container onClick={onClickMove}>
            <S.movieImage src={image} alt={title} />
            <S.movieTitle>{title}</S.movieTitle>
        </S.Container>
    );
};

export default Card;
