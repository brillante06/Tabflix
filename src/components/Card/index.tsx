import React, { FC, useRef } from 'react';
import moment from 'moment';
import { useHistory } from 'react-router';
import { useIntersecting } from '../../hooks/useIntersecting';
import { movieInfo } from '../../types';
import * as S from './styles';
import { AspectRatio, Image } from '../index';

interface movieCard {
    image: string;
    title: string;
    id: string;
    movie?: movieInfo;
    tag?: boolean;
}

const Card: FC<movieCard> = ({ image, title, id, movie, tag }) => {
    const history = useHistory();
    const lazyRef = useRef<HTMLImageElement | null>(null);
    const lazyLoading: IntersectionObserverCallback = (entries, observer) => {
        entries.forEach((entry) => {
            if (entry.isIntersecting) {
                const lazyimage = entry.target as HTMLImageElement;
                observer.unobserve(entry.target);
                if (lazyimage.dataset.src) lazyimage.src = lazyimage.dataset.src;
            }
        });
    };
    const onClick = () => {
        history.push(`/detail/${id}`);
    };
    useIntersecting(lazyRef, lazyLoading);
    return (
        <S.Container onClick={onClick}>
            <S.ImgWrapper>
                <AspectRatio ratio={16 / 10}>
                    <Image src={image} alt={title} ref={lazyRef} key={title} tag={tag} />
                </AspectRatio>
            </S.ImgWrapper>
            <S.InfoContainer>
                <S.MovieTitle>{title}</S.MovieTitle>
                <S.Info>
                    <S.MovieYear>{moment(movie?.release_date).year()}</S.MovieYear>
                    <S.MovieRating> ‚≠ê{movie?.vote_average}/10</S.MovieRating>
                </S.Info>
            </S.InfoContainer>
        </S.Container>
    );
};

export default React.memo(Card);
