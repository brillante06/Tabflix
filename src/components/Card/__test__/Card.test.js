import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';
import Card from '../index';
import { dummyMovie, dummyImage } from '../../../dummy/movieDummy';

describe('<Card />', () => {
    it('matches snapshot', () => {
        const onClick = jest.fn();
        const utils = render(
            <Card title="soul" image={dummyImage} onClick={onClick} movie={dummyMovie} />
        );
        expect(utils.container).toMatchSnapshot();
    });
    it('shows the props correctly', () => {
        const onClick = jest.fn();
        const utils = render(
            <Card title="soul" image={dummyImage} onClick={onClick} movie={dummyMovie} />
        );
        utils.getByText('soul');
        const image = utils.getByAltText('soul');
        expect(image).toHaveAttribute('src', dummyImage);
    });
});
