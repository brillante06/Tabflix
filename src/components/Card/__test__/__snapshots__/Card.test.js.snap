// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> matches snapshot 1`] = `
<div>
  <article
    class="sc-dlfnbm rBwJX"
  >
    <img
      alt="soul"
      class="sc-bdfBwQ dhhuN"
      src="https://image.tmdb.org/t/p/original/kf456ZqeC45XTvo6W9pW5clYKfQ.jpg"
    />
    <p
      class="sc-gsTCUz dookHH"
    >
      soul
    </p>
  </article>
</div>
`;
